### PR TITLE
fix: actor events: read entries from event index in order of their insertion

### PR DESCRIPTION
**Pull request by aarshkshah1992**
## Related Issues
Closes https://github.com/filecoin-project/lotus/issues/11823

## Proposed Changes
Event entries should be read from the event index in the order of their insertion

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/events/filter/index.go [hunk 1]
```diff
@@ -53,7 +53,8 @@ var ddls = []string{
 		flags BLOB NOT NULL,
 		key TEXT NOT NULL,
 		codec INTEGER,
-		value BLOB NOT NULL
+		value BLOB NOT NULL,
+		ordering INTEGER NOT NULL
```
Comment (Stebalien) [inline chain/events/filter/index.go:57]: nit: index?

File: chain/events/filter/index.go [hunk 2]
```diff
@@ -557,7 +602,7 @@ func (ei *EventIndex) prefillFilter(ctx context.Context, f *eventFilter, exclude
 		s = s + " WHERE " + strings.Join(clauses, " AND ")
 	}
 
-	s += " ORDER BY event.height DESC"
+	s += "ORDER BY event.height DESC"
```
Comment (Stebalien) [inline chain/events/filter/index.go:605]: I think we need the space, right?

File: chain/events/filter/index.go [hunk 3]
```diff
@@ -572,9 +617,27 @@ func (ei *EventIndex) prefillFilter(ctx context.Context, f *eventFilter, exclude
 		return xerrors.Errorf("exec prefill query: %w", err)
 	}
 
-	var ces []*CollectedEvent
+	// create struct for holding EventEntry with ordering
+	type eventEntryWithOrdering struct {
+		types.EventEntry
+		ordering uint64
```
Comment (Stebalien) [inline chain/events/filter/index.go:623]: I don't think we need to select this field, we just need to order by it when querying.

File: chain/events/filter/index.go [hunk 4]
```diff
@@ -672,8 +739,37 @@ func (ei *EventIndex) prefillFilter(ctx context.Context, f *eventFilter, exclude
...
 	// sort it into height order
-	sort.Slice(ces, func(i, j int) bool { return ces[i].Height < ces[j].Height })
-	f.setCollectedEvents(ces)
+	sort.Slice(ces, func(i, j int) bool { return ces[i].height < ces[j].height })
+
+	var collectedEvents []*CollectedEvent
+	// sort the entries in the collected event by ordering
+	for _, ce := range ces {
+		ce := ce
+		sort.Slice(ce.entries, func(i, j int) bool { return ce.entries[i].ordering < ce.entries[j].ordering })
```
Comment (Stebalien) [inline chain/events/filter/index.go:748]: Let's do this in the query.

File: chain/events/filter/index.go [hunk 5]
```diff
@@ -672,8 +739,37 @@ func (ei *EventIndex) prefillFilter(ctx context.Context, f *eventFilter, exclude
 
 	// collected event list is in inverted order since we selected only the most recent events
 	// sort it into height order
-	sort.Slice(ces, func(i, j int) bool { return ces[i].Height < ces[j].Height })
-	f.setCollectedEvents(ces)
+	sort.Slice(ces, func(i, j int) bool { return ces[i].height < ces[j].height })
```
Comment (Stebalien) [inline chain/events/filter/index.go:742]: I'm not sure why we bother with this at all...

File: chain/events/filter/index.go [hunk 6]
```diff
@@ -118,6 +120,35 @@ func (ei *EventIndex) initStatements() (err error) {
...
+	}
+	// rollback the transaction (a no-op if the transaction was already committed)
+	defer tx.Rollback() //nolint:errcheck
+
+	if _, err = tx.ExecContext(ctx, "ALTER TABLE event_entry ADD COLUMN ordering INTEGER NOT NULL DEFAULT 0"); err != nil {
+		return xerrors.Errorf("increment _meta version: %w", err)
+	}
+
+	// Add the index on event_id and ordering fields
+	if _, err = tx.ExecContext(ctx, "CREATE INDEX IF NOT EXISTS idx_event_id_ordering ON event_entry (event_id, ordering)"); err != nil {
```
Comment (Stebalien) [inline chain/events/filter/index.go:136]: We don't need to index on ordering, right?

File: chain/events/filter/index.go [hunk 7]
```diff
@@ -118,6 +120,35 @@ func (ei *EventIndex) initStatements() (err error) {
...
 
+func (ei *EventIndex) migrateToVersion3(ctx context.Context) error {
+	tx, err := ei.db.Begin()
+	if err != nil {
+		return xerrors.Errorf("begin transaction: %w", err)
+	}
+	// rollback the transaction (a no-op if the transaction was already committed)
+	defer tx.Rollback() //nolint:errcheck
+
+	if _, err = tx.ExecContext(ctx, "ALTER TABLE event_entry ADD COLUMN ordering INTEGER NOT NULL DEFAULT 0"); err != nil {
```
Comment (Stebalien) [inline chain/events/filter/index.go:131]: I'd prefer to actually fill these (may be possible in a single statement?).

File: chain/events/filter/index.go (hunk 3 above)
Comment (aarshkshah1992) [reply chain/events/filter/index.go:623]: @Stebalien For some reason, if I add this to the ordering, we get way more records than we want and the test starts failing. I spent time debugging it before giving up. 

I had changed the current ordering to  	s += "ORDER BY event.height DESC, event_entry.ordering ASC".

File: chain/events/filter/index.go (hunk 5 above)
Comment (aarshkshah1992) [reply chain/events/filter/index.go:742]: I agree. It was already here so didn't want to mess with it.

File: chain/events/filter/index.go (hunk 1 above)
Comment (aarshkshah1992) [reply chain/events/filter/index.go]: Done.

File: chain/events/filter/index.go (hunk 2 above)
Comment (aarshkshah1992) [reply chain/events/filter/index.go]: Done.

File: chain/events/filter/index.go (hunk 6 above)
Comment (aarshkshah1992) [reply chain/events/filter/index.go]: Done.

**Comment by aarshkshah1992** [comment]
This should be closed once https://github.com/filecoin-project/lotus/pull/11834 is merged.

**Pull request closed without merging**